File: python/ecat_types/free_species.py
import dataclasses
from dataclasses import dataclass

import molmass

from constants import ATOMIC_SYMBOLS, AVOGADRO, P_ATM_eVpA3, kB_eVpK


@dataclass(frozen=True, order=True)
class Token:
    symbol: str
    count: int


def make_token(symbol: str, count_str: str) -> Token:
    if not symbol or symbol not in ATOMIC_SYMBOLS:
        raise ValueError(f'Unknown atomic symbol {symbol}')

    count = int(count_str) if count_str else 1
    if count < 1:
        raise ValueError(f'Invalid count {count_str} for atomic symbol {symbol}')

    return Token(symbol, count)


def tokenize_molecule(molecule: str) -> list[Token]:
    tokens: list[Token] = []
    symbol = ''
    count_str = ''

    for char in molecule:
        if 'A' <= char <= 'Z':  # Start of a new atomic symbol
            if symbol:
                tokens.append(make_token(symbol, count_str))
            symbol = char
            count_str = ''
        elif symbol and 'a' <= char <= 'z':  # Part of an atomic symbol
            symbol += char
        elif symbol and '0' <= char <= '9':  # Part of a count
            count_str += char
        else:
            raise ValueError(f'Unexpected character {char} in molecule {molecule}')

    # Add the last token
    if symbol:
        tokens.append(make_token(symbol, count_str))

    return tokens


@dataclass(frozen=True, order=True)
class FreeSpecies:
    name: str
    phase: str
    _: dataclasses.KW_ONLY
    bond_length_A: float | None = None
    Etot_eV: float | None = None
    Efree_eV: float | None = None

    def __post_init__(self):
        if self.phase not in ['gas', 'aq', None]:
            raise ValueError(f'phase {self.phase} not supported. Species: {self.name}')

        if self.name == 'e-' and self.phase is not None:
            raise ValueError(f'e- should not have a phase, but {self.phase} was given')

        if self.name == 'H+' and self.phase != 'aq':
            raise ValueError('H+ can only exist in an aq phase')

        if self.name == 'OH-' and self.phase != 'aq':
            raise ValueError('OH- can only exist in an aq phase')

        if self.name in ['H2', 'O2'] and self.phase == 'aq':
            raise ValueError(f'aq solvation not supported for {self.name}')

    def __repr__(self) -> str:
        if self.phase is None:
            return self.name
        return f'{self.name}_{self.phase}'

    @property
    def is_charged(self) -> bool:
        return self.name[-1] in ['-', '+']

    @property
    def is_diatomic(self) -> bool:
        return self.is_homodiatomic or self.is_heterodiatomic

    @property
    def is_homodiatomic(self) -> bool:
        tokens = tokenize_molecule(self.name)
        return len(tokens) == 1 and tokens[0].count == 2

    @property
    def is_heterodiatomic(self) -> bool:
        tokens = tokenize_molecule(self.name)
        return len(tokens) == 2 and tokens[0].count == 1 and tokens[1].count == 1

    @property
    def molar_mass_gpMol(self):
        return molmass.Formula(self.name).mass

    def merge_into_dict(self, count: int, fs_dict: dict['FreeSpecies', int]) -> None:
        if self in fs_dict:
            fs_dict[self] += count
            if fs_dict[self] == 0:
                del fs_dict[self]
        else:
            fs_dict[self] = count

    def number_density_pA3(self, temperature_K: float) -> float:
        if self.phase == 'aq':
            if not hasattr(self, 'molar_mass_gpMol') or self.molar_mass_gpMol is None:
                raise ValueError(f"Molar mass of {str(self)} is not defined.")
            if self.name == 'H2O':
                density_gpcm3 = 0.997048  # www.wolframalpha.com/input?i=water+density
                density_gpA3 = density_gpcm3 * 1E-24
                return (density_gpA3 * AVOGADRO) / self.molar_mass_gpMol

        if self.phase == 'gas':
            pressure_eVpA3 = None
            vapor_pressure_atm = 1
            if self.name == 'O2':
                vapor_pressure_atm = 0.207  # 21 kPa, en.wikipedia.org/wiki/Vapor_pressure
                pressure_eVpA3 = vapor_pressure_atm * P_ATM_eVpA3
            elif self.name == 'H2':
                # The Standard Hydrogen Electrode (SHE) defines the pressure of H2 to be 1 atm
                pressure_eVpA3 = P_ATM_eVpA3
            elif self.name == 'H2O':
                vapor_pressure_atm = 0.0227  # 2.3 kPa, en.wikipedia.org/wiki/Vapor_pressure
                pressure_eVpA3 = vapor_pressure_atm * P_ATM_eVpA3
            else:
                raise ValueError(f"Vapor Pressure of {str(self)} has not yet been implemented.")

            return pressure_eVpA3 / (kB_eVpK * temperature_K)  # from ideal gas law: V/N = P/kT

        raise ValueError(f"Number Density of {str(self)} has not yet been implemented.")

File: von-react-app/src/EcatTypes/FreeSpecies.ts
import {
  AVOGADRO, ATOMIC_SYMBOLS_SET, P_ATM_eVpA3, kB_eVpK,
} from '../constants.ts';
import Calc from './Calc.ts';

interface Token {
  symbol: string,
  count: number,
}

function makeToken(symbol: string, countStr: string): Token {
  if (!ATOMIC_SYMBOLS_SET.has(symbol)) {
    throw new Error(`Unknown atomic symbol ${symbol}`);
  }

  const count = countStr ? parseInt(countStr, 10) : 1;
  if (Number.isNaN(count) || count < 1) {
    throw new Error(`Invalid count ${countStr} for atomic symbol ${symbol}`);
  }

  return { symbol, count };
}

/**
 * Tokenizes a molecule into its atomic components.
 * @param {string} molecule - The molecule string.
 * @returns {Token[]} An array of tokens.
 */
function tokenizeMolecule(molecule: string): Token[] {
  const tokens: Token[] = [];
  let symbol = '';
  let countStr = '';
  [...molecule].forEach((char) => {
    // If character is uppercase, it's the start of a new atomic symbol
    if (char >= 'A' && char <= 'Z') {
      if (symbol) {
        tokens.push(makeToken(symbol, countStr));
      }
      symbol = char;
      countStr = '';
    } else if (!symbol) {
      throw new Error(`Atomic symbol must start with uppercase letter, but got ${char} in molecule ${molecule}`);
    } else if (char >= 'a' && char <= 'z') {
      // If character is lowercase, it's part of an atomic symbol
      symbol += char;
    } else if (char >= '0' && char <= '9') {
      // If character is a digit, it's part of a count
      countStr += char;
    } else {
      throw new Error(`Unexpected character ${char} in molecule ${molecule}`);
    }
  });

  // Add the last token
  if (symbol) {
    tokens.push(makeToken(symbol, countStr));
  }

  return tokens;
}

export default class FreeSpecies {
  readonly name: string;
  readonly calc: Calc | null;
  readonly phase: string | null;
  readonly molar_mass_gpMol: number | null;

  constructor(name: string, phase: string | null, molar_mass_gpMol: number | null, calc: Calc | null) {
    this.name = name;
    this.calc = calc;
    this.phase = phase;
    this.molar_mass_gpMol = molar_mass_gpMol;
  }

  toString(): string {
    return this.phase ? `${this.name}_${this.phase}` : this.name;
  }

  /**
   * @return {boolean} Whether the species is a diatomic molecule.
   */
  get isDiatomic(): boolean {
    return this.isHomoDiatomic || this.isHeteroDiatomic;
  }

  /**
   * @return {boolean} Whether the species is a diatomic molecule of a single element.
   */
  get isHomoDiatomic(): boolean {
    const tokens = tokenizeMolecule(this.name);
    return tokens.length === 1 && tokens[0].count === 2;
  }

  /**
   * @return {boolean} Whether the species is a diatomic molecule of two different elements.
   */
  get isHeteroDiatomic(): boolean {
    const tokens = tokenizeMolecule(this.name);
    return tokens.length === 2 && tokens[0].count === 1 && tokens[1].count === 1;
  }

  /**
   * @param {number} temperature_K - The temperature in Kelvin.
   * @return {number} The number density in Angstrom^-3.
   */
  numberDensity_pA3(temperature_K: number): number {
    if (this.phase === 'aq') {
      if (!this.molar_mass_gpMol) {
        throw new Error(`Molar mass of ${String(this)} is not defined.`);
      }
      if (this.name === 'H2O') {
        const density_gpcm3 = 0.997048; // www.wolframalpha.com/input?i=water+density
        const density_gpA3 = density_gpcm3 * 1E-24;
        return (density_gpA3 * AVOGADRO) / this.molar_mass_gpMol;
      }
    }

    if (this.phase === 'gas') {
      let pressure_eVpA3;
      let vapor_pressure_atm = 1;
      if (this.name === 'O2') {
        vapor_pressure_atm = 0.207; // 21 kPa, en.wikipedia.org/wiki/Vapor_pressure
        pressure_eVpA3 = vapor_pressure_atm * P_ATM_eVpA3;
      } else if (this.name === 'H2') {
        // The Standard Hydrogen Electrode (SHE) defines the pressure of H2 to be 1 atm
        pressure_eVpA3 = P_ATM_eVpA3;
      } else if (this.name === 'H2O') {
        vapor_pressure_atm = 0.0227; // 2.3 kPa, en.wikipedia.org/wiki/Vapor_pressure
        pressure_eVpA3 = vapor_pressure_atm * P_ATM_eVpA3;
      } else {
        throw new Error(`Vapor Pressure of ${String(this)} has not yet been implemented.`);
      }

      return pressure_eVpA3 / (kB_eVpK * temperature_K); // from ideal gas law: V/N = P/kT
    }

    throw new Error(`Number Density of ${String(this)} has not yet been implemented.`);
  }
}
